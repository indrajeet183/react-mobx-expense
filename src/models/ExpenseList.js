import { types } from 'mobx-state-tree';

export const Expense = types.model({
    name:types.string,
    price:types.number,
    quantity:types.number,    
}).actions(self => ({
    changeName(newName) {
        self.name = newName
    },
    changePrice(newPrice) {
        self.price = newPrice
    },
    changeQuantity(newQuant) {
        self.quantity = newQuant
    },
})).views(self => ({
    get totalPrice () {
        return self.price * self.quantity
    }
}))

export const DailyExpense = types.model({
    date:types.string,
    //totalPrice:types.number,
    items:types.optional(types.array(Expense),[])
}).actions(self => ({
    changeDate (newDate) {
        self.date = newDate
    },    
    addItem (newItem) {
        self.items.push(newItem)
    },
    removeItem (name) {
        const index = self.items.findIndex(e => {
            if(e.name === name) 
                return e
        })
        self.items.splice(index,1)
    }
})).views(self => ({
    get totalPrice () {
        return self.items.reduce((sum,entry) => sum+(entry.price * entry.quantity),0)
    }
}))
