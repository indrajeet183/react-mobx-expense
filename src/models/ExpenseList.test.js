import {getSnapshot} from 'mobx-state-tree';

import {Expense, DailyExpense} from './ExpenseList';

const today = new Date ().toLocaleDateString ();

it ('can create instance of model Expense', () => {
  const expense = Expense.create ({
    name: 'First',
    price: 20,
    quantity: 4,
  });

  // expect (expense.name).toBe ('First');
  // expect (expense.price).toBe (20);
  // expect (expense.quantity).toBe (4);

  expect (getSnapshot (expense)).toMatchSnapshot ();
});

it ('can create instance of DailyExpense', () => {
  const dailyExpense = DailyExpense.create ({
    date: today,
    items: [
      {name: 'First Item', price: 20, quantity: 4},
      {name: 'Second Item', price: 10, quantity: 2},
    ],
  });

  // expect (dailyExpense.items.length).toBe (2);
  // expect (dailyExpense.totalPrice).toBe (100);
  // expect (dailyExpense.items[0].name).toBe ('First Item');
  // expect (dailyExpense.items[0].price).toBe (20);
  // expect (dailyExpense.items[0].quantity).toBe (4);

  expect (getSnapshot (dailyExpense)).toEqual ({
    date: today,
    items: [
      {name: 'First Item', price: 20, quantity: 4},
      {name: 'Second Item', price: 10, quantity: 2},
    ],
  });
});

it ('can add item in DailyExpense', () => {
  const dailyExpense = DailyExpense.create ({
    date: today,
    items: [],
  });

  dailyExpense.addItem ({name: 'First Item', price: 20, quantity: 4});

  expect (dailyExpense.items.length).toBe (1);
  expect (dailyExpense.items[0].name).toBe ('First Item');
  expect (dailyExpense.items[0].price).toBe (20);
  expect (dailyExpense.items[0].quantity).toBe (4);
});

it ('can remove item in DailyExpense', () => {
  const dailyExpense = DailyExpense.create ({
    date: today,
    items: [{name: 'First Item', price: 20, quantity: 4}],
  });

  expect (dailyExpense.items.length).toBe (1);

  dailyExpense.removeItem ('First Item');

  expect (dailyExpense.items.length).toBe (0);
});

it ('can calculate total price of DailyExpense', () => {
  const dailyExpense = DailyExpense.create ({
    date: today,    
    items: [
      {name: 'First Item', price: 20, quantity: 4},
      {name: 'Second Item', price: 10, quantity: 2},
      {name: 'Third Item', price: 5, quantity: 1}
    ],
  });

  expect(dailyExpense.totalPrice).toBe(105)
});
