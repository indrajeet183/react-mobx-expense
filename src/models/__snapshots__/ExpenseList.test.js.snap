// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create instance of model Expense 1`] = `
Object {
  "name": "First",
  "price": 20,
  "quantity": 4,
}
`;
